Node.js 코드 리뷰 요청드립니다.

SsureyMoon opened the pull request:
안녕하세요.
Node.js 로 웹 응용프로그램을 만들고 있는데, 노드를 사용해본지 오래되었고, 실무에서 노드를 써볼 기회가 없어서 잘 만들고 있는지 궁금해서 리뷰 요청을 드리게 되었습니다.

자세한 내용과 리뷰 질문 리스트는 https://github.com/SsureyMoon/code-review/blob/4353016038c45761904a6da28c4c05ac69daac4f/projects/koajs-proj-conf/README.md#%EB%A6%AC%EB%B7%B0-%EC%9A%94%EC%B2%AD 에 있습니다.

SsureyMoon committed:
[nodejs-review] init project

SsureyMoon committed:
[nodejs-review] init node.js app

SsureyMoon committed:
[nodejs-review] set up app and db container for dev

SsureyMoon committed:
[nodejs-review] integrate app with db and add testing healthcheck

SsureyMoon committed:
[nodejs-review] add book data update api and its test

SsureyMoon committed:
[nodejs-review] add an listing books api and its test

SsureyMoon committed:
[nodejs-review] add nodemon and code lint

SsureyMoon committed:
[nodejs-review] add review questions

miconblog commented on projects/koajs-proj-conf/src/routes.js:
```diff
@@ -0,0 +1,19 @@
+const router = new Router({
+    prefix: `/${config.API_VER}`,
+});
+router.get('/health', controllers.healthcheck);
```
저 헬쓰체크는 AWS Elb가 헬쓰를 던지기 위함인가요? 싱글쓰레드인 Node 앱을 직접 퍼블릭하게 오픈하면 DDos 취약합니다. 노드 앞에 Nginx 붙이고 Nginx에서 Node로 health를 요청할때 x-forward-xxx 형태의 헤더를 파싱해서 바로 200 상태코드만 반환해도 됩니다.

SsureyMoon replied:
답변주신 것을 잘 이해하지 못해서 추가 질문합니다. ㅠ
헤더를 파싱하는 곳이 Nginx인지 Node.js인지 궁금합니다.
health check 로 요청이 들어올때 Nginx 에서 x-forward-for을 파싱해서 가 ELB에서 온 경우에는 트래픽을 Node.js로 넘기지 않고 Nginx에서 바로 200을 리턴하면 되는 것인가요?

임시적으로 아래와 같이 Nginx 컨테이너를 추가하고 헬스체크 URL은 nginx에서 바로 200을 반환하도록 설정해주었습니다.
https://github.com/studiopluto/code-review/pull/31/commits/fe8d6516e47fff33b28365673e30958afb3ee4fe

miconblog replied:
헬스는 당연히 노드가 살아있는지 체크해야하니까 노드에서 해야하구요. 아마 ELB 에서 헬스체킹할때 포워딩되는 헤더가 있을꺼에요. 그 헤더만 보고 노드에서 바로 200으로 반환하면 어떨까 싶었습니다. 

검색하다가 노드 딴에서 명시적인 헬스라우팅 포인트를 두지않고 프록시로 넘어오는 헬쓰체크 헤더만 보고 반환하는게 괜찮아 보이더라구요. 링크는 나중에 찾으면 붙여드릴께요

SsureyMoon replied:
https://github.com/studiopluto/code-review/pull/31/commits/04faf443a9ea721d73a761a1a2744ef8a35173e4 로 수정하긴 했는데 맞게 했는지 확인 부탁드립니다 ㅠㅠ

자료를 찾아봤는데 없어서 EC2에 띄워서 시험해보았고, user-agent로 `ELB-HealthChecker/2.0`이 넘어오는 것을 확인했습니다. https://github.com/studiopluto/code-review/pull/31/commits/04faf443a9ea721d73a761a1a2744ef8a35173e4#diff-34d9a5435008de73db4c32f90b8cddf2R157

Node app의 healthcheck 컨트롤러와 엔드포인트를 삭제하고 https://github.com/studiopluto/code-review/pull/31/commits/04faf443a9ea721d73a761a1a2744ef8a35173e4#diff-f21894bb7e14a1c6ebdb838bf5e7cb98L15

미들웨어에서 라우팅 전에 모든 요청을 검사하면서 user-agent의 소문자 변환값이 `elb-healthchecker` 문자열을 포함하면 바로 200을 리턴해주도록 구현했습니다. https://github.com/studiopluto/code-review/pull/31/commits/04faf443a9ea721d73a761a1a2744ef8a35173e4#diff-ba7355b17465484a3b1621b02faeb3b4R24

miconblog replied:
헬스체크를 곰곰히 생각해보니까 이걸 어떻게 구현하느냐는 헬스체크를 하는 이유와 맞물리네요. 이것도 역시 방송때 질의하면서 다시 점검해봐요!

miconblog reviewed (commented):
수많은 프레임워크중에 Koa를 선택한 이유가 있을까요? async/await 가 나이스해보이지만 거의모든 어싱크 코드에 예외처리가 빠져있습니다.

miconblog commented on projects/koajs-proj-conf/README.md:
````diff
@@ -0,0 +1,77 @@
+        ├── database.test.js
+        └── middlewares.test.js
+```
+- 전체적인 프로젝트 구조가 맞는지 알고 싶습니다. (너무 범위가 넗다면 config 쪽과 db client 를 만들어 사용하는 부분, 미들웨어를 만드는 부분에 대해서만이라고 알려주시면 감사드리겠습니다.)
````
Koa 프레임워크는 저도 안써봐서 애매한데 best practice koa structure 라는 키워드로 검색해보세요. 

감이 잘 안오시면 express 프레임워크의 제너레이터가 생성해주는 구조를 한번 보세요

Node 앱은 src 폴더 하위에 모두 위치하고 실제 포트를 열고 listen 하는 코드는 root/bin 폴더에 www 라는 파일이름으로 쓰기도 합니다.

outsideris replied:
프로젝트 구조는 뭐 정답은 없는데 제가 보통 익숙한 구조에서는 `src`밑에 `test`가 있는건 약간 어색합니다. 보통은 `src`와 `test`가 같은 계층에 있다고 할까요? 

예전에는 그렇게 했는데 하다보니 이 관례도 약간 java에서 온 관례같아서 최근에는 `controller`, `service` `util` 등 용도별로 폴더를 만들고 `test`만 따로 분리해서 사용하다가 더 최근에는 아예 `test` 폴더를 안두고 소스 파일 옆에 같은 파일명에 테스트 파일. 그러니까 `serializers.js`가 있으면 바로 옆에 `serializers.test.js`를 두는 식도 많이 하고 있습니다.

SsureyMoon replied:
테스트 코드는 https://github.com/studiopluto/code-review/pull/31/commits/cc744425bbdb860f524af7848aca53b2aa70d20b 처럼 수정했고,

전체적인 프로젝트 구조는 https://github.com/studiopluto/code-review/pull/31/commits/cb956f1c306cdf50247a915e2981d1d09ff18040 처럼 수정했습니다.

SsureyMoon committed:
[nodejs-review] review-1 fixed. add exception handling via middleware

SsureyMoon committed:
[nodejs-review] review-2 half-fixed. add nginx proxy before node.js server

SsureyMoon commented:
수많은 프레임워크중에 Koa를 선택한 이유가 있을까요?
- async/await 를 잘 지원해주어서 선택하게 되었습니다.

거의 모든 async 코드에 예외처리가 빠져있습니다.
- [Koa.js 가이드](https://github.com/koajs/koa/wiki/Error-Handling)에서 제시한 것 처럼 https://github.com/studiopluto/code-review/pull/31/commits/d2295be9255d0108e4391b15eeec5d6930cd566f#diff-ba7355b17465484a3b1621b02faeb3b4R10 와 같이 미들웨어를 추가했고, NODE_ENV=production 인 경우 별도의 에러 로깅을 추가했습니다.
- await 구문마다 예외처리를 두는 것이 좋을까요?

miconblog commented on projects/koajs-proj-conf/README.md:
```diff
@@ -0,0 +1,80 @@
+│   └── ...
+├── package-lock.json
+├── package.json
+├── server.js
```
41번의 앱과 49번의 서버는 어떤 차이가 있죠? 제가 지금 지하철이라 소스 보기가 쉽지 않아서 일단 질문을 던져봅니다.

miconblog replied:
아하! app.js 는 코아앱이고 server.js는 포트붙여서 실제 띄우는 코드군요! 

저라면 app.js는 src/ 폴더 안으로 넣겠습니다. 코아앱이 여러개라면 src/ 라는 이름 대신에 apps/ 라는 이름을두고 다시 폴더로 app1/ app2/ 이런 형태로 관리합니다. 

그리고 실제 포트를 붙여서 서버를 띄우는 server.js 코드는 root/bin/ 폴더 하위에 둡니다.

SsureyMoon replied:
https://github.com/studiopluto/code-review/pull/31/commits/cb956f1c306cdf50247a915e2981d1d09ff18040#diff-0c789222a79ea6d00958b212a194eca1 처럼 프로젝트 루트 밑에 bin/www 로 옮겼습니다.

miconblog commented on projects/koajs-proj-conf/README.md:
```diff
@@ -0,0 +1,80 @@
+    ├── db
+    │   ├── client.js
+    │   └── index.js
+    ├── config.js
```
설정파일은 src 밖으로 빼고, 실제 시크릿 정보를 담은 형태면 어떨까요? 

커밋은 config.sample.js 혹은 json 형태로 네이밍해서 샘플만 커밋하는 형태로~

SsureyMoon replied:
https://github.com/studiopluto/code-review/pull/31/commits/bee3c2acefacfeed99ba81146b54b9fb692ab00b 로 수정했습니다.
config 를 src 밖으로 뺐습니다.

샘플 파일을 만들고 파일에 실제 시크릿 정보를 담으면, 여러 인스턴스로 배포할 때 환경변수보다 관리하기 힘들 것 같다고 생각해서, 환경변수로 시크릿 정보를 세팅하는 부분은 변경하지 않았습니다.

miconblog commented on projects/koajs-proj-conf/README.md:
````diff
@@ -0,0 +1,80 @@
+        └── middlewares.test.js
+```
+- 전체적인 프로젝트 구조가 맞는지 알고 싶습니다. (너무 범위가 넗다면 config 쪽과 db client 를 만들어 사용하는 부분, 미들웨어를 만드는 부분에 대해서만이라고 알려주시면 감사드리겠습니다.)
+- ./app.js 에서 database client를 ctx에 세팅해서 컨트롤러에서 사용하고 있습니다.(./app.js#L18) 이런 방법이 올바른 사용 방법인지 알고 싶습니다.
````
코아는 안써봤지만 쓰신 형태를 보니 그래프큐엘 서버처럼 컨텍스트로 의존성 객체를 주입해주는 형태인가 보군요. 맞게 쓰신거 같은데,.. 요건 제가 Koa를 좀 살펴보고 알려드릴께요.

outsideris replied:
저도 koa를 잘 모르긴 한데 graphql을 안써봐서 그런지 몰라도 미들웨어로 db를 앱단에서 콘텍스트에 넣는 것은 저한테는 어색하게 느껴집니다. 
node.js에 상관없이 설계를 하면 레이어링을 하게 되는데 디비 레이어가 앱의 최상단인 `app.js`까지 올라온 이유를 잘 모르겠습니다. 각 컨트롤러 혹은 서비스에서 `import`해서 사용해도 충분하지 않은가요?

miconblog commented on projects/koajs-proj-conf/README.md:
```diff
@@ -0,0 +1,80 @@
+
+##### 테스트 방법
+- 현재는 호출되는 다른 함수들을 모킹해서 테스트하고 (./src/tests/controllers.test.js#L40) DB에 실제 값을 써봐야 할 경우 테스트용 컬렉션을 생성해서(./src/config.js#L13) 테스트하고 테스트후에 지우고 있습니다.
+같은 DB 내에서 개발용/테스트용 컬렉션을 나누어 쓰게되는데 문제가 없는지 궁금합니다.
```
로컬에서 테스트를 돌릴때 개발용 정보를 오염시키지만 않으면 상관없을것 같아요.

outsideris replied:
나누어 쓰는 방식은 루비등에서 많이 사용하는 방식이긴 합니다. 겹치지 않으니까 로컬에선 오히려 편하고 프로덕션에 나갈일은 없으니까요.

제 취향에서는 `config.js`에 테스트용 내용을 하드코딩하기 보다는 어차피 환경변수로 작성되어 있으므로 [여기](https://github.com/SsureyMoon/code-review/blob/review/nodejs-newbie/projects/koajs-proj-conf/package.json#L9)에서 환경변수로 테스트 컬렉션을 설정할 것 같습니다.

SsureyMoon replied:
https://github.com/studiopluto/code-review/pull/31/commits/160e03aea864ef23dd5ef0d3df62dcdcfbc09a7f 로 수정했습니다.

miconblog commented on projects/koajs-proj-conf/README.md:
```diff
@@ -0,0 +1,80 @@
+- 현재는 호출되는 다른 함수들을 모킹해서 테스트하고 (./src/tests/controllers.test.js#L40) DB에 실제 값을 써봐야 할 경우 테스트용 컬렉션을 생성해서(./src/config.js#L13) 테스트하고 테스트후에 지우고 있습니다.
+같은 DB 내에서 개발용/테스트용 컬렉션을 나누어 쓰게되는데 문제가 없는지 궁금합니다.
+
+- Node.js 에서 사용하는 테스팅 패턴이 있다면 알려주시면 감사드리겠습니다.
```
요건 @outsideris 아싸님이 더 잘 아실듯.

테스트 주도 개발로 작성하는 테스트 코드와 코드를 작성하고 방어하는 테스트 코드의 성격은 조금 다르구요 

API 단위로 태스트하는 통합테스트와 펑션단위의 유닛테스트도 성격이 다릅니다. 

테스트 성격에따라서 패턴도 좀 다르고 파일 관리도 달라집니다.

outsideris replied:
테스팅 패턴이 노드라고 아주 다르진 않습니다. 질문이 너무 커서 어디서부터 얘기해야할지를 잘 모르겠습니다.

miconblog commented on projects/koajs-proj-conf/README.md:
```diff
@@ -0,0 +1,80 @@
+- Node.js 에서 사용하는 테스팅 패턴이 있다면 알려주시면 감사드리겠습니다.
+
+#### Node.js 앱 프로덕션으로 실행 방법
+Dockerfile에서 `node ./server.js` 으로 노드 앱을 실행하고 있습니다. (./Dockerfile#L24) 컨테이너 기반환경에서 Node.js 서버를 동작시키는 방법을 추천해주시면 감사드리겠습니다.
```
node ./bin/www 형태로 실행해주세요. 아마 찾아보면 권장 패턴이 있을꺼에요.

outsideris replied:
노드에서 가장 일반적인 방법은 `npm start` 입니다. 거의 관례적이라 어느 노드앱이라도 `npm start`를 하면 뭔가 프로덕션용이 시작될거라고 기대합니다. [여기](https://github.com/SsureyMoon/code-review/blob/review/nodejs-newbie/projects/koajs-proj-conf/package.json#L8)를 보면 `npm start`를 개발용으로 쓰고 있는데 `nodemon`을 쓰는건 `dev`로 만들어서 `npm run dev`로 실행하시는 편이 좋아보입니다.

SsureyMoon replied:
찾아보니 pm2 라는 것도 있던데 컨테이너 환경에서는 잘 사용하지 않나요?

SsureyMoon replied:
https://github.com/studiopluto/code-review/pull/31/commits/cb956f1c306cdf50247a915e2981d1d09ff18040 로 수정했습니다.

outsideris replied:
보통 node.js는 watch dog(모니터링하다가 죽으면 다시 살리는)을 쓰지만 Container에서는 demon으로 띄울 이유도 없고 컨테이너가 그 역할을 해야하므로 따로 pm2 등을 쓰지 않습니다.(최소한 저는...)

Edunga1 replied:
안녕하세요! 저는 공식 문서 https://nodejs.org/ko/docs/guides/nodejs-docker-webapp/#dockerfile 에서 `npm start` 라고 보긴 했는데 이 문서에서 링크한 docker-node best practice라는 문서는 `node` 커맨드를 사용한다고 하네요.
https://github.com/nodejs/docker-node/blob/master/docs/BestPractices.md#cmd

miconblog replied:
저도 위에 저 문서와 비슷한 맥락으로 npm 명령어대신 node로 스크립트를 직접 실행하라는 얘기를 어디선가 본거 같아요. 하도 많은 문서를 봐서 ㅎㅎㅎ 확신이 없었는데,... ㅋ  감사합니다. 

제가 본 문서는 도커에서 npm start 로 했을때 뭔가 문제 상황이 발생해서  node bin/www 로 해결했다는 문서였는데 스택오바플로우 답변인지 헷갈리네요.

miconblog replied:
초창기엔 pm2 말고 forever 라는 모듈로 죽은 노드앱들을 다시 살리는 용도로 데몬을 많이 띄웠어요. pm2도 여전히 같은 용도인데, 개발모드에선 저도 안쓰구요. 라이브에서도 써볼기회가 아직 없었네요. 써보신분들 경험이 궁금합니다.

outsideris replied:
읽어보지 않은 문서인데 감사합니다.

SsureyMoon replied:
https://github.com/studiopluto/code-review/pull/31/commits/bd156169169fd6223af3328eb082bbf7ffa5ffa7 로 수정했습니다. 감사합니다~!

miconblog commented on projects/koajs-proj-conf/README.md:
```diff
@@ -0,0 +1,80 @@
+Dockerfile에서 `node ./server.js` 으로 노드 앱을 실행하고 있습니다. (./Dockerfile#L24) 컨테이너 기반환경에서 Node.js 서버를 동작시키는 방법을 추천해주시면 감사드리겠습니다.
+
+#### 기타
+- Sanitizing, Validation, Serialze 를 직접 구현해서 사용했는데, 다른 추천하는 방법이나 라이브러리 등이 있으면 알려주시면 감사드리겠습니다.
```
직접 구현하면 구멍이 많아서 전 그냥 ORM 씁니다.

outsideris replied:
저도 가져다쓰는 편입니다. node.js 는 유닉스 철학을 많이 따르는 편이라 별 거지같은 라이브러리도 다 npm으로 있는 편이라서 가져다가 사용합니다. 각 용도별로 찾기 어려우시면 따로 얘기해주시면 제가 주로 쓰는걸 추천해 드리겠습니다.

SsureyMoon replied:
사용하고 계시는 라이브러리가 있다면 추천해주시면 감사하겠습니다.

miconblog replied:
전 parse-server 좋아합니다~ :) Koa 랑 궁합이 맞는지는 모르겠어요.

outsideris commented:
아하! 리뷰가 올라왔는데 몰랐군요

outsideris commented on projects/koajs-proj-conf/app.js:
```diff
@@ -0,0 +1,36 @@
+    transports: [
+        new winston.transports.File({ filename: 'error.log', level: 'error' }),
+    ],
+});
```
docker도 쓰고 있어서 파일로 쓰기보다는 stderr로 내리는게 더 낫지 않나요?

SsureyMoon replied:
https://github.com/studiopluto/code-review/pull/31/commits/f96c66d4e375a8a87dc7e9f795f5d7ef98a857f3 로 수정했습니다.

outsideris commented on projects/koajs-proj-conf/package.json:
```diff
@@ -0,0 +1,34 @@
+  "main": "app.js",
+  "scripts": {
+    "lint": "npx eslint ./app.js ./server.js ./src",
+    "start": "npx nodemon ./server.js",
```
노드에서 npm 모듈을 설치하면 `node_modules/.bin/`아래 실행파일이 생깁니다. 여기에 `eslint`나 `nodemon` 등이 있는데 지금 여기서 npm script를 사용할때는(`npm start`나 `npm run lint` 같은) 자동으로 이곳을 패스로 잡아주므로 npx를 쓰지 마시고 `nodemon ./server.js`로 해도 됩니다.

npx는 create-react-app 처럼 npm으로 설치하지 않고 명령어를 실행하려고 할때 사용하는데 여기서는 이미 설치하셨으므로 이렇게 하시면 오히려 더 오래걸리겁니다.

SsureyMoon replied:
https://github.com/studiopluto/code-review/pull/31/commits/f5002fc343b81ffd97f0b708f748b7e53182347e 로 수정했습니다.

outsideris commented on projects/koajs-proj-conf/server.js:
```diff
@@ -0,0 +1,10 @@
+
+const app = require('./app');
+
+
```
아마 파이썬이 이렇게 두줄을 띄울꺼고 노드는 두줄을 띄우지 않습니다. ㅋㅋ

SsureyMoon replied:
https://github.com/studiopluto/code-review/pull/31/commits/8df8fd5432c4758092326e53044dd533509fa187  로 수정했습니다.

outsideris commented on projects/koajs-proj-conf/src/db/client.js:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    async healthcheck() {
+        await this.collection.get();
```
이렇게 안써봐서 헷갈리는데 이렇게 하면 `get()` 한 값이 리턴되나요?

SsureyMoon replied:
그냥 컬렉션 정보가 리턴됩니다. 헬스체크시에 db에 정상적으로 접근되는지 테스트 하기 위해 추가해주었습니다.
https://docs.arangodb.com/devel/Drivers/JS/Reference/Collection/#collectionget

SsureyMoon replied:
제가 질문을 잘못 이해했군요. return 없이 await 만 써도 리턴됩니다.

miconblog replied:
jest.fn() 의 반환 값은 프로미스가 아니라 그냥 함수 아닌가요? async/await 없이 써야할것 같아요. 
근데 의도는 왠지 이해가 됩니다. DB 컬렉션 조회 자체가 비동기라서 async/await를 붙여놓은거 같네요. 의도하신거죠?

SsureyMoon replied:
넵. DB 컬렉션 조회가 비동기라 await를 붙여주었습니다.

그리고 대부분의 테스트코드에 jest.fn()이 사용되었는데, 프로미스를 반환하지 않은 것은 실수였습니다.;;
일반 함수가  await와 잘 동작하는게 신기해서 찾아봤는데 아래와 같은 논의를 발견했습니다.
https://esdiscuss.org/topic/await-on-synchronous-functions#content-14

그래서 지금 테스트 코드를 궂이 프로미스를 반환하는 형태로 수정하지 않아도 될것 같습니다.

miconblog replied:
await 가 일반 함수에도 붙일 수 있지만 코드를 읽는 사람의 입장에는 비동기로 동작한다라는 전제가 붙어서 당연히 프로미스가 넘어오겠거니 생각할수도 있어서 그 의식의 흐름에 따라 어? 이건 프로미스가 아닌데? 했다가 그럼 뭐지? 따라가보니까 jest.fn()이 보였고, 어? 그럼 테스트 의도가 도대체 뭘까? 라는 질문까지 왔습니다. 

코드를 보다가 테스트 하려는 목적에 맞게 jest.fn() 을 쓴건가? 라는 의문이 생기더라구요. 그래서 이 질문은 글로 하기 어려우니 방송때 직접 여쭤보겠습니다~ ㅋ

outsideris commented on projects/koajs-proj-conf/src/db/index.js:
```diff
@@ -0,0 +1,32 @@
+const arangojs = require('arangojs');
+
+const dbConfig = require('../config').database;
```
이거 그냥 아래처럼 하면 될것 같습니다. 
```
const { database: dbConfig } = require('../config');
```

SsureyMoon replied:
https://github.com/studiopluto/code-review/pull/31/commits/694be07cdd0b24fea05156c641272f19832eae29#diff-b1a322f871a836f8caf44a9889f98b7eR3 로 수정했습니다.

outsideris commented on projects/koajs-proj-conf/src/db/index.js:
```diff
@@ -0,0 +1,32 @@
+    }, () => {
+        console.log('collection access failed'); // eslint-disable-line
+    });
+});
```
`then`에 `then`이 중첩되는건 좋지 않고 예외처리도 좋지 앖습니다. `then`에 `then`을 체인으로 연결해서 `collection.create()`를 하시면 됩니다.

SsureyMoon replied:
https://github.com/studiopluto/code-review/pull/31/commits/694be07cdd0b24fea05156c641272f19832eae29#diff-b1a322f871a836f8caf44a9889f98b7eR20 처럼 수정했습니다.

outsideris commented on projects/koajs-proj-conf/src/db/index.js:
```diff
@@ -0,0 +1,32 @@
+module.exports = {
+    db,
+    collection,
+    client: new DatabaseClient(db, collection),
```
쓰신 디비 라이브러리가 제가 안쓰는거라서 헷갈리는데 `exports`를 하면 다른 곳에서 `require`로 가져다 쓰는데 이때 `require`는 sync 함수입니다. 위에서 보면 `collection.get()`가 있는데 이함수의 정확한 역할을 모르겠지만 `require`는 sync로 동작하고 `collection.get()`의 결과는 async로 돌아오기 때문에 타이밍 문제가 생길수도 있겠다 싶습니다.(안해봤습니다.)

SsureyMoon replied:
넵. 타이밍 문제가 생길 수 있을 것 같습니다 ㅠㅠ 처음에 앱이 실행될때 필요할 컬렉션이 없으면 만들어주기 위해 추가 해주었고, 최초 한번(컬렉션이 없을 때)가 아니면 문제가 발생할 가능성이 없어서 발견 못한 것 같습니다.

이 코드에서 컬렉션을 생성하는 부분을 제거하고, 별도의 초기화 작업으로 컬렉션 생성을 추가해서, 앱 실행시는 컬렉션이 존재한다고 가정하는 것은 어떨까요?

outsideris replied:
이런 부분은 초기에는 잘 발견안되는데 개발하면서 코드가 복잡해 지면 타이밍 문제가 생기고 발견되기 어렵곤 합니다. 이렇게 디비 초기화하는 부분은 node.js에서 종종 문제가 되고는 하는데 앱 실행시 최초에 불러오고 초기화되어서 약간의 시간이 지난뒤에도 값이 채워지면 크게 상관은 없습니다. 하지만 여기서는 `new DatabaseClient(db, collection)` 부분이 타이밍 문제가 생기면 의도와 다르게 `new DatabaseClient(db, null),`로 들어갈 수 있다고 쓰다보니까...
`const collection = db.collection(dbConfig.dbCollection)` 부분이 있어서 그런일은 안생길수도 있겠네요. `collection()`의  내부를 봐야 타이밍 문제가 생길지 아닐지 않수 있을것 같습니다.

outsideris commented on projects/koajs-proj-conf/src/middlewares.js:
```diff
@@ -0,0 +1,44 @@
+    const properties = property.split('.');
+    return properties.reduce((prevObj, key) => {
+        return prevObj[key];
+    }, obj);
```
```
prevObj, key) => {
        return prevObj[key];
}
```

같은 경우는 

```
(prevObj, key) => prevObj[key]
```

로 써도 되지 않나 싶네요.

outsideris replied:
근데 이거 정확히 뭐하는 함수죠? 값만 가져오시려는 건가요?

SsureyMoon replied:
Koajs 에서 URL 쿼리는 request.query 로 넘어오고, POST body는 request.body 로 넘어옵니다.

두개다 같은 validation 로직을 활용할 수 있을 것 같아서, https://github.com/studiopluto/code-review/pull/31/files#diff-34d9a5435008de73db4c32f90b8cddf2R9 처럼, GET일 때와 POST일때 다른 오브젝트를 파싱하도록 하기 위해서 위와 같은 함수를 만들었습니다.

SsureyMoon replied:
https://github.com/studiopluto/code-review/pull/31/commits/592063b842fca3ba35e316b43de2da22c742b6e5 로 수정했습니다.

outsideris commented on projects/koajs-proj-conf/src/serializers.js:
```diff
@@ -0,0 +1,35 @@
+    return {
+        convert: (obj, key) => ({ [idField]: obj[key] }),
+    };
+};
```
```
const buildFieldSerialzier = idField => ({
    convert: (obj, key) => ({ [idField]: obj[key] }),
});
```

이렇게 쓰며 어떤가요? 

근데 `covert`는 어디서 쓰나요?

SsureyMoon replied:
https://github.com/studiopluto/code-review/pull/31/commits/592063b842fca3ba35e316b43de2da22c742b6e5 로 수정했습니다.

book 객체를 응답용으로 시리얼라이즈할때, 필드 이름을 변경해주기 위해서 추가해주었습니다. https://github.com/studiopluto/code-review/pull/31/files#diff-85005188e0617f0896ee158ce329d049R10 처럼 _key 필드의 이름을 id로 바꾸어 주기 위해 사용되었습니다.

outsideris commented on projects/koajs-proj-conf/src/serializers.js:
```diff
@@ -0,0 +1,35 @@
+};
+
+const serialize = (serializersToUse, objToSerialize) => {
+    return Object.keys(objToSerialize).reduce((prevObj, key) => {
```
keys를 가져와서 reduce를 하는 코드가 여러번 보이는데.... 어떤 게 필요해서 이렇게 하시는건가요? 제가 잘 안하던 패턴인데 실행해 보진 않아서요.

SsureyMoon replied:
https://github.com/studiopluto/code-review/pull/31/files/fe8d6516e47fff33b28365673e30958afb3ee4fe#diff-85005188e0617f0896ee158ce329d049R11 처럼 응답 생성시 시리얼라이저를 만들어두고, 값 변환+필터링(시리얼라이즈 할 오브젝트의 키에 해당하는 시리얼라이저가 있으면 값변환, 없으면 해당 필드 제거)를 해주기 위해서 만들었습니다.

위와 마찬가지로, 요청 바디를 검증시 Validator를 만들어 두고 값 검증+필터링을 해주는 부분에도 reduce가 사용되었습니다.

outsideris commented on projects/koajs-proj-conf/src/tests/apit.test.js:
```diff
@@ -0,0 +1,101 @@
+const db = require('../db');
+const { API_VER } = require('../config');
+
+describe('healthcheck endpoint', async () => {
```
describe는 `async` 필요가 없어보입니다.

SsureyMoon replied:
https://github.com/studiopluto/code-review/pull/31/commits/4633b5f6f539eb0fedc8a5e42c9d682dc34b4586 로 수정했습니다.

outsideris reviewed (commented):
일단 이정도만 봤습니다...

SsureyMoon committed:
[nodejs-review] review-3 fixed. move test code to src dir from test dir

SsureyMoon committed:
[nodejs-review] review-4 fixed. refactor project structure
- move app.js in src
- change server.js to bin/www executable
- npm start for prod, npm run dev for develop environment

SsureyMoon committed:
[nodejs-review] review-5 fixed. remove redundant async for describe

SsureyMoon committed:
[nodejs-review] review-6 fixed. remove two line breaks

SsureyMoon committed:
[nodejs-review] review-7 fixed. remove redundant npx in npm script

SsureyMoon committed:
[nodejs-review] review-8 fixed. refactor db/index.js
- refactor importing database object
- remove a then callback in another then callback

SsureyMoon commented:
꼼꼼하게 리뷰해주셔서 정말 감사합니다!
일단 쉽게 수정할 수 있는 것들부터 수정해서 커밋했고, 나머지도 빠르게 수정해서 커밋 추가하겠습니다.

SsureyMoon committed:
[nodejs-review] review-2 fully-fixed. handle healthcheck from elb in a middleware
- remove healthcheck handling in nginx
- remove healthcheck endpoint in node app
- check every traffic whether it has user-agent header set by ELB
- return 200 OK when a traffic is from ELB

SsureyMoon committed:
[nodejs-review] review-9 fixed. shorten object statments

SsureyMoon committed:
[nodejs-review] review-10 fixed. move config file out of src dir

SsureyMoon committed:
[nodejs-review] review-11 fixed. set the test collection in npm script

miconblog commented on projects/koajs-proj-conf/init/db-init.sh:
```diff
@@ -0,0 +1,26 @@
+arangosh \
```
와우! 아랭고라는 DB도 있었군요. 이 DB를 선택한 이유가 궁금하네요. 어떤 장점이 있을까요?

SsureyMoon replied:
graphql 지원이 되고 성능이 좋다고 들어서, 이번 기회에 배워보고 싶어서 선택했습니다.

miconblog replied:
성능이 좋다라는 벤치마크 자료는 아랭고 홈페이지에서 확인할수있는데, graphql 을 지원한다는 얘기는 없는데요? ㅎㅎㅎ graphql을 지원한다는 의미와 graph query를 효율적으로 처리한다는 의미는 다릅니다. 

그리프큐엘을 지원한다면 prisma 에서 지원한다는 의미로 저는 해석하는데,.. 역시나 아랭고 디비는 아직이네요. 
https://www.prisma.io/features/databases/ 

고로 좋은 선택을 하셨는지는 잘 모르겠습니다. ^^ 참고~

SsureyMoon replied:
graphql이 아니라 graphdb 지원인데 잘못 적었습니다;;

miconblog commented on projects/koajs-proj-conf/init/db-init.sh:
```diff
@@ -0,0 +1,26 @@
+const dbname='\"$DB_NAME\"';
+
+const createUser = (username, passwd, dbname, userDao) => {
+    if(userDao.all().map(obj=>obj.user).indexOf(username) < 0){
```
아랭고는 Dao 라는 표현을 자주 쓰나요? 

근데 쉘 파일 안에서 자바스크립트를 써도 되나요? @,@ 대 혼돈~ ㅋ

SsureyMoon replied:
Dao는 마땅히 붙여줄 이름이 없어서 사용했습니다ㅠ 사용되지 않는 용어인듯 합니다.

쉘파일 안의 자바스크립트 코드는 그냥 문자열입니다 ㅎ
--javascript.execute-string 옵션으로 자바스크립트 코드를 문자열로 보내 arangosh 내에서 초기화 구문을 실행시키기 위해 추가해주었습니다.

--javascript.execute 옵션으로 자바스크립트 파일을 실행할 수도 있다고 하는데, 별도로 뺄 경우 환경변수를 어떻게 읽어야 할지 애매해서 저런식으로 쉘 + 자바스크립트의 혼합 파일이 생성되었습니다;;

miconblog replied:
어떤 의도인지는 이해했는데,... 음.... 더 나은 대안을 한번 고민해보죠~

miconblog commented on projects/koajs-proj-conf/src/db/client.js:
```diff
@@ -0,0 +1,28 @@
+            RETURN r
+        `);
+        const results = await booksCursor.map(value => serializers.serializeBook(value));
+        return results || [];
```
async/await 구분에 try/catch가 없으면 db가 실제로 커넥션 되지 않았거나 운용중에 끊겼을때 응답을 못받는 불상사가 생길수있습니다. 에러처리 해주세요!

miconblog replied:
Koa의 에러처리를 이해하고 보니까 여기서 안해도 코아 미들웨어딴에서 예외를 잡아줄것만 같네요. 

테스트하나만 해주세요. 디비 연결하지 않고 요 구문에서 에러날때 에러핸들러가 반응하는지 확인 좀 ,... ~ 제가 더 궁금하네요~ ㅋ

SsureyMoon replied:
개발 서버에서 테스트 했을때는 예외가 발생할때 미들웨어에서 잡아주는 것을 확인했습니다.
디비 연결을 끊고 books URL을 호출했을때 에러핸들러로 들어오는 것을 자동화 테스트로 만들지는 못했습니다.(어떻게 만드는지 감이 안잡혀서..)

대신 에러 핸들러 미들웨어의 next 콜백에서 임의로 에러를 발생시키는 테스트는 아래와 같이 만들었습니다.
https://github.com/SsureyMoon/code-review/commit/d2295be9255d0108e4391b15eeec5d6930cd566f#diff-6a8c6e65d42218ccece3d3503fda2373R88

miconblog replied:
아... 테스트를 작성해달란 얘긴 아니엇는데,... ㅋㅋ 그냥 디비서버 실행안하고 해보면 어떻게 캐치되나 궁금했습니다

miconblog commented on projects/koajs-proj-conf/src/db/index.js:
```diff
@@ -0,0 +1,36 @@
+    })
+    .catch(() => {
+        throw new Error('DB connection failed');
+    });
```
와~ 캐치 캐치도 연달아 되는군요. @,@ 

http://dexie.org/docs/Promise/Promise.catch()

outsideris replied:
저도 처음봤네요. 근데 불남님이 주신 링크는 타입으로 제한해서 캐치하는데 여기선 그냥 함수만 있는데 이래도 연달아 되나요?

miconblog replied:
ㅎㄷㄷ 저도 연속 캐치는 안써봤어요. 문님이 저렇게 써서 돌려봤을테니까 되는거겠죠?

miconblog replied:
근데 의도는 알겠는데 건강한(?) 방법인지는 모르겠습니다. 실제 의도는 DB에 컬렉션을 가져와서 없으면 새로 만드는게 의도인데, 캐치에 줄줄이 걸어놓은걸 보면 일단, DB 조회할때 컬렉션이 없으면 catch가 되는지부터 의문이네요. 

커넥션이 없으면 당연히 catch로 걸릴텐데 커넥션이 없어서 캐치 했는데 거기다가 바로 create()를 하면 안되는거죠. 이게 일단 이상한 점이고,.. 

왠지 컬렉션이 없는 경우에 빈값을 갖는 then 을 탈꺼 같습니다. 문님 확인 좀 해주세요.. ㅋ

miconblog replied:
그리고 의도한대로라면 일단 첫번째 캐치에서 DB 커넥션 에러가 아닌경우에만 create를 하는게 어떨까 싶네요.

SsureyMoon replied:
컬렉션이 없는 경우에는 reject 되어서 두번째 catch 구문을 타고, 두번째 구문에서 collection.create()에서 예외가 났을때 세번째 catch 로 넘어갑니다.

SsureyMoon replied:
https://github.com/studiopluto/code-review/pull/31/commits/d3d882f3a469549f110fe7bc61a6e7006309032c 로 수정했습니다.
error가 ArangoError인 경우(데이터베이스 접근 가능), 에러코드가 1203(collection not found)인 경우에만 컬렉션을 생성해주도록 했습니다.

outsideris commented on projects/koajs-proj-conf/src/api.test.js:
```diff
@@ -0,0 +1,109 @@
+    afterEach(async () => {
+        // delete all docs in the collection after each test
+        await db.collection.truncate();
+    });
```
이런 훅이 반복되는데 같은 파일의 경우는 전체를 하나의 `describe`로 묶어 주시고 정의하시면 반복해서 쓰지 않아도 됩니다.

SsureyMoon replied:
https://github.com/studiopluto/code-review/pull/31/commits/146e6cbefb8ab68a14cfaaa92f19422ca091f04c 로 수정했습니다.

outsideris commented on projects/koajs-proj-conf/src/middlewares.js:
```diff
@@ -0,0 +1,52 @@
+exports.setContext = contextObj => async (ctx, next) => {
+    Object.keys(contextObj).forEach((key) => {
+        ctx[key] = contextObj[key];
+    });
```
이 코드 보니까 `contextObj`의 모든 키의 값은 `ctx` 객체의 같은 값이 있으면 덮어쓰고 엎으면 추가하시려는 것 같은데 그냥 `Object.assign`쓰시면 될것 같습니다.

```
> var o1 = {a:1, b:2}
undefined
> var o2 = {b:3, c:4}
undefined
> Object.assign(o1, o2)
{ a: 1, b: 3, c: 4 }
```

miconblog replied:
middleware.setContext() 함수가 코아에서 제공해주는 방법인줄알고 문서를 한참 봤는데,... ㅎㅎㅎ 직접 만든거였다니,.... 허탈~ ㅋㅋㅋ 어쩐지 좀 이상하다 싶었습니다 ㅋ

요거 지우시고 app.context.db=db() 형태로 app.js에서 데이터베이스 주입해주세요.

SsureyMoon replied:
처음 개발할때 생각은 conte 에 db 외에 다른 것들도 주입해줘야 할 것 같아서 setContext 미들웨어를 만들었었습니다. 말씀하신대로 바로 주입해주는것이 좋을 것 같아서 https://github.com/studiopluto/code-review/pull/31/commits/dc69a2d2c7c82acd14483baed37e74c3dc4bbf33 로 수정했습니다.

outsideris commented on projects/koajs-proj-conf/src/middlewares.js:
```diff
@@ -0,0 +1,52 @@
+const resolve = (obj, property) => {
+    const properties = property.split('.');
+    return properties.reduce((prevObj, key) => prevObj[key], obj);
+};
```
`resolve`의 함수명이 너무 범용적이라서 그런지 잘 이해가 안되네요.

SsureyMoon replied:
https://github.com/studiopluto/code-review/pull/31/commits/2d7f65f96a6540d0820d358041ec919fdff7eb44 로 수정했습니다.

outsideris commented on projects/koajs-proj-conf/src/routes.js:
```diff
@@ -0,0 +1,18 @@
+    prefix: `/${config.API_VER}`,
+});
+router.get('/books', controllers.listBooks);
+router.post('/books', buildValidateMiddleware(bookItemValidators, 'request.body'), controllers.postBook);
```
`'request.body'`에서 의도는 이해했습니다. 아까 커멘트 봤을때는 잘 이해 못했는데요. 개인적으로는 꼭 필요한 경우아니면 이런 문자열은 실수를 찾아내기 어려워서 선호하는 편은 아닙니다. koa의 방식을 잘 모르겠긴 한데 각 라우터가 미들웨어에서 컨텍스트로 잡아준 `ctx`의 의존적이라서 이런 접근을 하게 된건 아닌가 생각해보게 되네요.

outsideris commented on projects/koajs-proj-conf/src/validators.js:
```diff
@@ -0,0 +1,69 @@
+            };
+        }
+        return { ...prevObj };
+    }, {});
```
실제 로직은 완전히 이해하려면 좀 더 테스트해봐야겠지만 `reduce`를 쓰는데 클로저로 외부의 `sanitizersObj`를 사용하는건 좋은 접근은 아닌것 같습니다.

miconblog commented on projects/koajs-proj-conf/app.js:
```diff
@@ -0,0 +1,36 @@
+    });
+}
+
+app.use(middlewares.handleExceptions);
```
에러처리 관련해서 코아 문서를 보니까 미들웨어 처리가 순서대로 스택에 쌓았다가 하나씩 미들웨어를 처리하는 구조더라구요. 그래서 에러처리는 첫번째 미들웨어로 등록하라고 하는데, 위에 보면 로거가 에러핸들러보다 먼저 꼽혀있어서 개발모드에선 에러가 발생할때 로깅이 안될것 같아요.  혹시 의도 하신건가요?

SsureyMoon replied:
koa-logger 문서를 봤을때 가장 위에 등록해주라고 그렇게 등록해주었습니다. 말씀하신대로 로깅이 안되어야 정상인데 에러가 발생할 때 계속 콘솔에 찍히는데 그 이유를 찾아봤더니
https://github.com/SsureyMoon/code-review/blob/review/nodejs-newbie/projects/koajs-proj-conf/src/middlewares.js#L17 에서 emit 해줘서 그런 것 같습니다.

SsureyMoon replied:
https://github.com/studiopluto/code-review/pull/31/commits/71c0b143e83b61c663ef485f53126aa34839a8e6 로 수정했습니다.

miconblog commented on projects/koajs-proj-conf/src/app.js:
```diff
@@ -0,0 +1,37 @@
+    database: {
+        client: db.client,
+    },
+}));
```
중간에 미들웨어에 데이터베이스를 주입한 특별한 이유가 있을까요? 코어 공식 문서 app.context 섹션을 보면 그냥 상단에 app.context.db = db.client 형식으로 써도 될것 같은데,... 

코어가 요청을 처리하는 구조상 저렇게 중간에 미들웨어로 꼽는다고해서 큰 이점이 있어보이진 않습니다. 제가 이해한게 맞다면,..  ^^;;; 

특별한 이유가 없다면 코드를 좀 더 간결하게 쓰는게 어떨까 싶네요.

miconblog replied:
참고로 Koa 에선 database 를 실제쓰이는 컨트롤러에서 require 구문으로 땡기지 않고 컨텍스트에 주입하는게 권장 패턴이라고 코아 문서에 써 있네요. @outsideris 

코아는 반복적으로 require 쓰는게 맘에 들지 않았나봐요~ 그래서 컨텍스트 공유 객체를 만든것 같아요.

miconblog replied:
그리고 ctx.database.client.xxxxx 형태로 쓰는건 좀 길지 않나요? 
ctx.db.xxxxx 형태로 쓰는게 코드 읽기는 더 좋은듯

SsureyMoon replied:
https://github.com/studiopluto/code-review/pull/31/commits/7e7e6e7b060495fc6781e6317d2a5fe51dc14889 로 수정했습니다.

miconblog reviewed (commented):
역시 지하철 리뷰는 힘들지만 덕분에 Koa 공부를 엄청 했네요~ 문님 감사!

SsureyMoon committed:
[nodejs-review] review-12 fixed. try creating collection when db is connected

SsureyMoon committed:
[nodejs-review] review-13 fixed. log errors to stderr instead of file

SsureyMoon committed:
[nodejs-review] review-14 fixed. move error handling middleware to the top

SsureyMoon committed:
[nodejs-review] review-15 fixed. shorten ctx.database.client.xx to ctx.db.xx

SsureyMoon committed:
[nodejs-review] review-16 fixed. remove setContext, inject db object directly

SsureyMoon committed:
[nodejs-review] review-17 fixed. change the name of function resolve to extractProperty

SsureyMoon committed:
[nodejs-review] review-18 fixed. run app on container via native node cmd

SsureyMoon committed:
[nodejs-review] review-19 fixed. use common setup and teardown in higher describe level

outsideris commented:
https://www.youtube.com/watch?v=DvavoYFpmpE
